common/types/string.go

```diff
@@ -221,3 +228,101 @@ func stringStartsWith(s String, pre ref.Val) ref.Val {
 	}
 	return Bool(strings.HasPrefix(string(s), string(preStr)))
 }
+
+func stringFormat(s String, args ref.Val) ref.Val {
+	formatStr := string(s)
+	argList, ok := args.(traits.Lister)
+	if !ok {
+		return MaybeNoSuchOverloadErr(argList)
+	}
+	i := 0
+	var builtStr strings.Builder
+	for i < len(formatStr) {
+		if formatStr[i] == '{' {
+			if i+1 < len(formatStr) && formatStr[i+1] == '{' {
+				err := builtStr.WriteByte('{')
+				if err != nil {
+					return NewErr("error writing format string: %s", err.Error())
+				}
+				i += 2
+				continue
+			}
+			newIndex, val, refErr := parseFormatArg(i, formatStr, argList)
+			if refErr != nil {
+				return refErr
+			}
+			_, err := builtStr.WriteString(val)
+			if err != nil {
+				return NewErr("error writing format string: %s", err.Error())
+			}
+			i = newIndex
+		} else if formatStr[i] == '}' {
+			if i+1 < len(formatStr) && formatStr[i+1] == '}' {
+				err := builtStr.WriteByte('}')
+				if err != nil {
+					return NewErr("error writing format string: %s", err.Error())
+				}
+				i += 2
+				continue
+			} else {
+				return NewErr("unmatched '}'")
+			}
+		} else {
+			err := builtStr.WriteByte(formatStr[i])
+			if err != nil {
+				return NewErr("error writing format string: %s", err.Error())
+			}
+			i++
+		}
+	}
+	return String(builtStr.String())
+}
+
+func parseFormatArg(lastIndex int, formatStr string, args traits.Lister) (int, string, ref.Val) {
+	i := lastIndex + 1
+
+	i, argIndex, err := parseArgIndex(i, formatStr)
+	if err != nil {
+		return -1, "", NewErr("error while parsing arg index: %s", err.Error())
+	}
+	refSize := args.Size()
+	argSize, ok := refSize.(Int)
+	if !ok {
+		return -1, "", NewErr("couldn't convert arg size")
+	}
+	if Int(argIndex) >= argSize {
+		return -1, "", NewErr("index %d out of range", argIndex)
+	}
+
+	argAny := args.Get(Int(argIndex))
+	argStrVal := argAny.ConvertToType(StringType)
+	argStr, ok := argStrVal.Value().(string)
+	if !ok {
+		return -1, "", NewErr("string conversion failed")
+	}
+	return i, argStr, nil
+}
+
+func parseArgIndex(lastIndex int, formatStr string) (int, int, error) {
+	i := lastIndex
+	var indexStr strings.Builder
+	for true {
+		if i >= len(formatStr) {
+			return -1, -1, errors.New("could not find matching '}'")
+		} else if formatStr[i] == '}' {
+			if indexStr.Len() == 0 {
+				return -1, -1, errors.New("no argument index given (\"{}\" is not valid)")
+			}
+			break
+		} else if !unicode.IsDigit(rune(formatStr[i])) {
+			return -1, -1, fmt.Errorf("unrecognized character '%c' found in argument index", formatStr[i])
+		}
+		indexStr.WriteByte(formatStr[i])
+		i++
+	}
+	argIndex, err := strconv.Atoi(indexStr.String())
```

Comment:
Is `strconv.Atoi` compatible with all the digit representations `unicode.IsDigit` supports? Do you want to support just ASCII digits?

Reply:
`Atoi` only works on ASCII digits (so does `ParseInt`, consequently), so at least as written it just works on ASCII digits. For what it's worth, Python handles Bengali here, I think:

```
>> "{৩}".format(1, 2, 3, 4, 5)
'4'
>>> "{১}".format(1, 2, 3, 4, 5)
'2'
>>> "{১৪}".format(1, 2, 3, 4,)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: Replacement index 14 out of range for positional args tuple
```

It's worth pointing out that I picked other systems from [this list](https://en.wikipedia.org/wiki/List_of_numeral_systems#By_culture_/_time_period) and tried them as well, but couldn't get any to work.

I also tried a similar snippet with C#:

```c#
decimal temp = 20.4m;
string s = String.Format("The temperature is {০}°C.", temp);
Console.WriteLine(s);
```

But that gave me an error:

```
System.FormatException: Input string was not in a correct format.
```

(If the index was parsed, it would have resulted in `System.FormatException: Index (zero based) must be greater than or equal to zero and less than the size of the argument list.`)

Based on all this, I think it'd be safest to keep CEL support to just ASCII digits (that seems to be the common denominator across languages/string formatters here), but I'm curious what others think.

Reply:
Oh, I'm all in favor of ascii only. My point was that you use Unicode.IsDigit above and maybe should swap that out with something that only matches ascii digits?

Reply:
Right, I'll mark this as resolved once I do.